src/rust/aggregator/src/data/leaderboards.rs

```diff
@@ -0,0 +1,293 @@
+    }
+
+    fn poll_interval(&self) -> Option<std::time::Duration> {
+        Some(std::time::Duration::from_secs(5))
+    }
+
+    async fn process_and_save_internal(&mut self) -> DataAggregationResult {
+        let mut transaction = self
+            .pool
+            .begin()
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        let competitions = sqlx::query_as!(
+            Competition,
+            r#"
+                SELECT * FROM competition_metadata
+                WHERE start < CURRENT_TIMESTAMP AND CURRENT_TIMESTAMP < "end"
+            "#,
+        )
+        .fetch_all(&mut transaction as &mut PgConnection)
+        .await
+        .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        for comp in competitions {
+            aggregate_data_for_competition(&mut transaction, comp).await?;
+        }
+        transaction
+            .commit()
+            .await
+            .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        Ok(())
+    }
+}
+
+async fn aggregate_data_for_competition<'a>(
+    transaction: &mut Transaction<'a, Postgres>,
+    comp: Competition,
+) -> DataAggregationResult {
+    sqlx::query!(
+        r#"
+            WITH homogenous_fills AS (
+                SELECT DISTINCT taker_address AS "user"
+                FROM fill_events
+                WHERE time > $2 AND time < $3
+                AND NOT EXISTS (
+                    SELECT * FROM competition_indexed_events
+                    WHERE fill_events.txn_version = competition_indexed_events.txn_version
+                    AND fill_events.event_idx = competition_indexed_events.event_idx
+                    AND competition_indexed_events.competition_id = $1
+                )
+                UNION
+                SELECT DISTINCT maker_address AS "user"
+                FROM fill_events
+                WHERE time > $2 AND time < $3
+                AND NOT EXISTS (
+                    SELECT * FROM competition_indexed_events
+                    WHERE fill_events.txn_version = competition_indexed_events.txn_version
+                    AND fill_events.event_idx = competition_indexed_events.event_idx
+                    AND competition_indexed_events.competition_id = $1
+                )
+            )
+            INSERT INTO competition_leaderboard_users ("user", volume, integrators_used, n_trades ,competition_id)
+            SELECT "user", 0, '{}'::text[], 0, $1 FROM homogenous_fills WHERE "user" NOT IN (
+                SELECT "user" FROM competition_leaderboard_users WHERE competition_id = $1
+            )
+        "#,
+        comp.id,
+        comp.start,
+        comp.end,
+    )
+    .execute(transaction as &mut PgConnection)
+    .await
+    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+    let users = sqlx::query!(
+        r#"
+            SELECT * FROM competition_leaderboard_users WHERE "user" IN (
```

Comment:
Probably want a SELECT DISTINCT here and immediately below.

Reply:
This changed.